## Peak calling
### 1. MACS2
In our calculations, those narrow peaks with a p-value equal to 0.01 were selected by MACS2 or [MACS3](https://github.com/macs3-project/MACS)
```shell
macs3 callpeak -t Sample.bam -c Control.bam -f BAMPE -g hs -n Sample --outdir peak_consider_background -method p 0.01 --bdg 2> outputSummary
```
If no background bam file available, peaks can also be called with single input
```shell
macs3 callpeak -t Sample.bam -f BAMPE -g hs -n Sample --outdir peak_without_background -method p 0.01 --bdg 2> outputSummary
```

Besides, it is also possible to perform advanced step-by-step peak calling. In the official website for [MACS3](https://macs3-project.github.io/MACS/index.html), you can also find an advanced step-by-step peak calling tutorial, which can be combined with our case if you want to consider spike-In when doing peak calling

```shell
#!/bin/sh

# determine global parameters
genome_size=2700000000

# we use sample 1 and its control sample 15 as an example
i=1
Sample=Sample_$i
if [ $i -lt 5 ]
then
    j=$(echo $i+14 | bc)
    Control=Sample_$j
else
    j=$(echo $i+10 | bc)
    Control=Sample_$j
fi
echo "Do peak calling for ${Sample} by considering background from ${Control}"
# determine individual parameters 
d=241
ctrl_num_read=12501765
treat_spike_num_read=99
ctrl_spike_num_read=37235

macs3 pileup -f BAMPE -i ${Sample}.sorted.rmDup.rmBL.bam -B --extsize $d -o ${Sample}.pileup.bdg
## build local background track from control
# d background
d2=$(echo $d/2 | bc)
macs3 pileup -f BAMPE -i ${Control}.sorted.rmDup.rmBL.bam -B --extsize $d2 -o ${Control}_d_bg.pileup.bdg
# slocal background
d3=$(echo "scale=4;"$d/1000 | bc)
macs3 pileup -f BAMPE -i ${Control}.sorted.rmDup.rmBL.bam -B --extsize 500 -o ${Control}_1k_bg.pileup.bdg
macs3 bdgopt -i ${Control}_1k_bg.pileup.bdg -m multiply -p $d3 -o ${Control}_1k_bg_norm.bdg
# llocal background
d4=$(echo "scale=4;"$d/10000 | bc)
macs3 pileup -f BAMPE -i ${Control}.sorted.rmDup.rmBL.bam -B --extsize 5000 -o ${Control}_10k_bg.pileup.bdg
macs3 bdgopt -i ${Control}_10k_bg.pileup.bdg -m multiply -p $d4 -o ${Control}_10k_bg_norm.bdg
# Combine and generate the maximum background noise
macs3 bdgcmp -m max -t ${Control}_1k_bg_norm.bdg -c ${Control}_10k_bg_norm.bdg -o ${Control}_1k_10k_bg_norm.bdg
macs3 bdgcmp -m max -t ${Control}_1k_10k_bg_norm.bdg -c ${Control}_d_bg.pileup.bdg -o ${Control}_d_1k_10k_bg_norm.bdg
## scale the ChIP and control to the same sequencing depth
# p=ctrl_num_read*d/genome_size
p=$(echo "scale=8;"$ctrl_num_read\*$d/$genome_size | bc)
macs3 bdgopt -i ${Control}_d_1k_10k_bg_norm.bdg -m max -p $p -o ${Control}_local_bias_raw.bdg
# scaling treat and control based on spike-in
# q=ctrl_spike_num_read/treat_spike_num_read
q=$(echo "scale=8;"$treat_spike_num_read/$ctrl_spike_num_read | bc)
macs3 bdgopt -i ${Control}_local_bias_raw.bdg -m multiply -p $q -o ${Control}_local_lambda.bdg
macs3 bdgopt -i ${Sample}.pileup.bdg -m multiply -p 1 -o ${Sample}_scale.pileup.bdg
## compare ChIP and local lambda to get the scores in qvalue
macs3 bdgcmp -t ${Sample}_scale.pileup.bdg -c ${Control}_local_lambda.bdg -m qpois -o ${Sample}_qvalue.bdg
# or pvalue
macs3 bdgcmp -t ${Sample}_scale.pileup.bdg -c ${Control}_local_lambda.bdg -m ppois -o ${Sample}_pvalue.bdg
# significant peaks with a cutoff c=2, which means the p value should be smaller than 10e-2
macs3 bdgpeakcall -i ${Sample}_pvalue.bdg -c 2 -l $d -o ${Sample}_qval_peaks.bed

```

### 2. SEACR

SEACR accept bedgraph files when doing peak calling. Firstly, The `sam` file can be transformed to `bam` and `bed` file with samtools and bedtools for further processing

sam >> bam >> bed

```shell
samtools view -bS -F 0x04 alignment/sam/bowtie2.sorted.rmDup.sam >alignment/bam/bowtie2.mapped.bam

bedtools bamtobed -i alignment/bam/bowtie2.mapped.bam -bedpe >alignment/bed/bowtie2.bed

awk '$1==$4 && $6-$2 < 1000 {print $0}' alignment/bed/bowtie2.bed >alignment/bed/bowtie2.clean.bed

cut -f 1,2,6 alignment/bed/bowtie2.clean.bed | sort -k1,1 -k2,2n -k3,3n  >alignment/bed/bowtie2.fragments.bed
```

Next, the bed files can be spike-In normalized to bedgraph files with `bedtools genomecov`

```shell
# export the sequence depth from E.coli `sam` file
seqDepthDouble = `samtools view -F 0x04 alignment/sam/bowtie2_spikeIn.sam | wc -l`
seqDepth = $((seqDepthDouble/2))
echo $seqDepth > alignment/sam/bowtie2_summary/bowtie2_spikeIn.seqDepth
# calculate scale factor of spike-In
scale_factor=`echo "10000 / $seqDepth" | bc -l`
# normalize bed file to bedgraph file
bedtools genomecov -bg -scale $scale_factor -i alignment/bed/bowtie2.fragments.bed -g alignment/hg38.chrom.sizes > alignment/bedgraph/bowtie2.fragments.normalized.bedgraph
```

Now, it's time to performing peak calling with SEACR

[SEACR](https://github.com/FredHutch/SEACR) is a tool for calling peaks and enriched regions from sparse CUT&RUN or chromatin profiling data. Before applying SEACR, `SEACR_1.3.sh` and `SEACR_1.3.R` should be downloaded first.

```shell
bash SEACR_1.3.sh target.bedgraph control.bedgraph non relaxed output.peaks
```
Here, `control.bedgraph` are used as the background for `target.bedgraph`. Since we have normalized fragment counts with the E.coli read count, we set the normalization option of SEACR to `non`. Otherwise, the `norm` is recommended. 




